com.unity.netcode.gameobjects/Runtime/Core/NetworkManager.cs

```diff
@@ -2123,6 +2154,44 @@ private void SyncTime()
 #endif
         }
 
+        internal struct DisconnectReasonMessage : INetworkMessage
+        {
+            public string Reason;
+
+            public void Serialize(FastBufferWriter writer)
+            {
+                if (writer.TryBeginWrite(sizeof(int) + FastBufferWriter.GetWriteSize(Reason)))
+                {
+                    writer.WriteValue(Reason.Length);
+                    writer.WriteValue(Reason);
+                }
+                else
+                {
+                    writer.TryBeginWrite(sizeof(int) + FastBufferWriter.GetWriteSize(""));
+                    writer.WriteValue(0);
+                    writer.WriteValue("");
+                    Debug.LogWarning("Disconnect reason didn't fit. Disconnected without sending a reason. Consider shortening the reason string.");
+                }
+            }
+
+            public bool Deserialize(FastBufferReader reader, ref NetworkContext context)
+            {
+                int size;
+                reader.TryBeginRead(sizeof(int));
+                reader.ReadValue(out size);
+                Reason = new string(' ', size);
```

Comment:
Strings are immutable - this allocation will immediately be thrown away... can you just avoid it?

Reply:
Yes, this part is painful to look at. But that's really a limitation of the messaging system. I need a way to tell it "I'll be reading a string of length x". And the only idiomatic way to do so, in this system, is to pass a string of length x to GetWriteSize(). Hey, I'd be happy with `reader.UnsafeReadValue` or if we could have implicit `TryBeginRead` with throw-on-exceptions, but we don't. 

So, short of refactoring the MessagingSystem, taking a one-string-alloc upon being disconnected seemed pretty far on the minor side of the spectrum.

Reply:
But thanks for pointing out the Read/Write**Safe** part, this is now unnecessary